zkevm-circuits/src/evm_circuit/execution/shl_shr.rs

```diff
@@ -59,38 +60,50 @@ impl<F: Field> ExecutionGadget<F> for ShlShrGadget<F> {
         let is_shl = OpcodeId::SHR.expr() - opcode.expr();
         let is_shr = 1.expr() - is_shl.expr();
 
-        let quotient = cb.query_word_rlc();
-        let divisor = cb.query_word_rlc();
-        let remainder = cb.query_word_rlc();
-        let dividend = cb.query_word_rlc();
-        let shift = cb.query_word_rlc();
+        let quotient = cb.query_word32();
+        let divisor = cb.query_word32();
+        let remainder = cb.query_word32();
+        let dividend = cb.query_word32();
+        let shift = cb.query_word32();
         let shf0 = cb.query_cell();
 
         let mul_add_words =
             MulAddWordsGadget::construct(cb, [&quotient, &divisor, &remainder, &dividend]);
-        let shf_lt256 = IsZeroGadget::construct(cb, sum::expr(&shift.cells[1..32]));
-        let divisor_is_zero = IsZeroGadget::construct(cb, sum::expr(&divisor.cells));
-        let remainder_is_zero = IsZeroGadget::construct(cb, sum::expr(&remainder.cells));
-        let remainder_lt_divisor = LtWordGadget::construct(cb, &remainder, &divisor);
+        let shf_lt256 = IsZeroGadget::construct(cb, sum::expr(&shift.limbs[1..32]));
+        let divisor_is_zero = IsZeroWordGadget::construct(cb, &divisor);
+        let remainder_is_zero = IsZeroWordGadget::construct(cb, &remainder);
+        let remainder_lt_divisor =
+            LtWordGadget::construct(cb, &remainder.to_word(), &divisor.to_word());
 
         // Constrain stack pops and pushes as:
         // - for SHL, two pops are shift and quotient, and push is dividend.
         // - for SHR, two pops are shift and dividend, and push is quotient.
-        cb.stack_pop(shift.expr());
-        cb.stack_pop(is_shl.expr() * quotient.expr() + is_shr.expr() * dividend.expr());
-        cb.stack_push(
-            (is_shl.expr() * dividend.expr() + is_shr.expr() * quotient.expr())
-                * (1.expr() - divisor_is_zero.expr()),
+        cb.stack_pop_word(shift.to_word());
+        cb.stack_pop_word(
+            quotient
+                .to_word()
+                .mul_selector(is_shl.expr())
+                .add_unchecked(dividend.to_word().mul_selector(is_shr.expr())),
```

Comment:
I think we could find a better way to write these kind of expressions (I also saw them in another opcode).

So originally we had:
```rust
v = is_shl.expr() * dividend.expr() + is_shr.expr() * quotient.expr())
                * (1.expr() - divisor_is_zero.expr()
```

The logic for this is the following:
```rust
v = if (not (divisor_is_zero) {
  if (is_shl) { 
    dividend 
  } else if (is_shr) {
    quotient 
  } else {
    0
  }
} else {
  0
};
```

So maybe we could have some helpers to write these if/else where the conditions are booleans and the values are words.  But let's leave this for a future PR!  I just wanted to share my thoughts here :)

Reply:
Yes, agree current unchecked style looks very ugly and hard to read 🥲 
It will be good to left it as a future PR and check whether can leverage `macro!` features to make it looks confortable!